### DISCUSSION TITLE
How i can add image into pdf with my custom button?

### DISCUSSION THREAD

**Discussion by mrbeku** [stephanrauh/ngx-extended-pdf-viewer#1862] [Q&A]
How i can add image into pdf with my custom button? For example add QR code image into PDF document

**Comment by stephanrauh** [comment]
Honest answer: I don't know. But I can give you some pointer.

First of all, do you want to modify the PDF file or do you want to add the image to the annotation layer? From a user's point of view, there's probably little to no difference. However, technically, the annotation layer is just an add-on to the PDF file.

If you want to modify the core PDF file, you need a PDF manipulation library. There are many suitable libraries out there, both for node.js and for the most popular server-side languages. I don't know if you can manipulate the PDF file directly in the browser. Most architects consider this a bad practice, so I never examined this.

If you're happy with adding the image to the annotation layer, examine the source code of the stamp tool: https://github.com/stephanrauh/pdf.js/blob/bleeding-edge/src/display/editor/stamp.js

Sooner or later I'll have to implement such a feature in `NgxPdfViewerService.addEditorAnnotation()`. But I'm not there yet, and I doubt I'll manage to implement it this year. Can I talk you into examining the implementation of the stamp to and to sending me a pull request improving the `addEditorAnnotation()` method?

Best regards,
Stephan
